src/boxrobots/src/controllers/QPinnerloopcontroller.jl

```diff
@@ -0,0 +1,476 @@
+using PyCall
+@pyimport boxatlas
+@pyimport boxatlas.boxatlas as boxatlastypes
+
+abstract ContactSwitchType
+immutable NoSwitch <: ContactSwitchType end
+immutable MakingContact <: ContactSwitchType end
+immutable BreakingContact <: ContactSwitchType end
+
+type ContactSwitch{T<:ContactSwitchType}
+  limb_symbol::Symbol
+end
+
+function ContactSwitch{T}(limb_symbol::Symbol, ::Type{T})
+  return ContactSwitch{T}(limb_symbol)
+end
+
+function ContactSwitch(cs_0::ContactState, cs_1::ContactState)
+  """
+  Compute the contact switch type when transitioning from
+  cs_0 --> cs_1
+  """
+
+  num_contact_state_mismatches = 0
+  limb_symbol = Symbol()
+  contact_switch_type = ContactSwitchType
+
+  for (limb_sym, contact_indicator_0) in cs_0
+    contact_indicator_1 = cs_1[limb_sym]
+    # there was a contact switch
+    if contact_indicator_0 != contact_indicator_1
+      if (contact_indicator_0 & !contact_indicator_1)
+        contact_switch_type = BreakingContact
+      else
+        contact_switch_type = MakingContact
+      end
+      limb_symbol = limb_sym
+      num_contact_state_mismatches += 1
+    end
+  end
+
+  if num_contact_state_mismatches > 1
+    error("more than one state changed simultaneously, dont know how to handle this")
+  end
+
+  if num_contact_state_mismatches > 0
+    contact_switch = ContactSwitch(limb_symbol, contact_switch_type)
+    return true, contact_switch
+  else
+    return false, nothing
+  end
+
+end
+
+function get_contact_switch_type{T}(contact_switch::ContactSwitch{T})
+  return T
+end
+
+
+abstract ContactMismatchType
+immutable NoMismatch <: ContactMismatchType end
+immutable EarlyContact <: ContactMismatchType end
+immutable LateContact <: ContactMismatchType end
+
+type ContactMismatch{T<:ContactMismatchType}
+  limb_sym::Symbol
+end
+
+function ContactMismatch{T}(limb_symbol::Symbol, ::Type{T})
+  return ContactMismatch{T}(limb_symbol)
+end
+
+type QPInnerLoopControllerState
+  t_start::Float64 # time at which soln_data was constructed
+  soln_data::PyObject
+  contact_plan::ContactPlan
+  contact_state::ContactState
+  contact_mismatch::ContactMismatch
```

Comment:
Another performance tip: when possible, avoid incompletely defined types as your fields (ContactMismatch is not a concrete type, since it's missing its `{T}` parameter)

 https://docs.julialang.org/en/stable/manual/performance-tips/#avoid-fields-with-abstract-containers